dissect/target/helpers/hashutil.py

```diff
@@ -60,34 +62,73 @@ def custom(fh, algos):
     return _hash(fh, ctx)
 
 
-def hash_uri_records(target, record: Record) -> Record:
+def hash_uri_records(target: Target, record: Record) -> Record:
     """Hash uri paths inside the record."""
-    uri_fields = (field for field in record._field_types.items() if issubclass(field[1], path))
+    warnings.warn(
+        (
+            "The hash_uri_records() function is deprecated, and will be removed in dissect.target 3.15. ",
+            "Use hash_path_records() instead",
+        ),
+        DeprecationWarning,
+    )
+    return hash_path_records(target, record)
 
-    hashed_uri_records = []
-    for name, _ in uri_fields:
-        try:
-            uri_record = hash_uri(target, getattr(record, name))
-            hashed_uri_records.append(uri_record)
-        except (FileNotFoundError, IsADirectoryError):
-            pass
 
-    if not hashed_uri_records:
-        return record
+def hash_path_records(target: Target, record: Record) -> Record:
+    """Hash files from path fields inside the record."""
+    hashed_paths = []
 
     if target.os == "windows":
-        path_type = windows_path
+        path_type = fieldtypes.windows_path
     else:
-        path_type = posix_path
+        path_type = fieldtypes.posix_path
+
+    for field_name, field_type in record._field_types.items():
+        if issubclass(field_type, fieldtypes.path):
+            path = getattr(record, field_name, None)
+
+            if path is None:
+                continue
+
+            path = str(path)
+
+            try:
+                resolved_path = target.resolve(path)
+                path_hash = target.fs.hash(resolved_path)
+            except (FileNotFoundError, IsADirectoryError):
+                pass
+            else:
+                if path != resolved_path:
+                    warnings.warn(
+                        (
+                            "The function passing in a path field did not resolve that path. ",
+                            "The autmatic resolving of path fields will be removed in the future.",
```

Comment:
```suggestion
                            "The automatic resolving of path fields will be removed in the future.",
```

Reply:
Why, by the way?

Reply:
Because this part of the code should not have any knowledge about what the meaning of the content of a field is. That is the responsibility of the plugin creating the record. The plugin knows whether it is useful to resolve a path or not. It's correct separation of responsibilities.

Reply:
What about when for correctness I want to put the original path from e.g. the registry in the record which might include an environment variable, but want to hash that file? Should I create a second field like `real_path`?

Reply:
That is a good point. You do want to preserve original values. Though putting unresolved paths with e.g. env variables feels problematic. In my mind a (pathlib style) path type means a literal path, where percentages, dollars and all are part of the actual path, not of potential env variables, otherwise you would still be guessing if it is an actual path or if it still needs to be resolved.

In an ideal world I think you would have 2 values, a literal string which contains whatever was parsed from the artifact and a path which is resolved and present on the target. You would still have the side cases of: 1. what if the path correctly resolves but is not present in the target and 2. what if the path does not or partially resolve how can you be sure the dollar/percentage is part of the path or an unresolved env variable.

Although the last side case won't be solvable in practice, I do think we need to move to a system where a path is as fully resolved as possible and if the literal value, is of value it should be present in the record as a second (string) value. Also because people tend to want to put paths that come out of target-query into something like target-fs to get to the file, which will fail if it is not resolved (we could of course change target-fs to (have the option to) resolve paths).

But given that this is more a policy discussion then something that should be realized by this PR, I'll revert the deprecation warning on unresolved paths for now.